pandas/core/generic.py

```diff
@@ -2985,18 +2986,45 @@ def astype(self, dtype, copy=True, raise_on_error=True, **kwargs):
 
         Parameters
         ----------
-        dtype : numpy.dtype or Python type
+        dtype : data type, or dict of column name -> data type
+            Use a numpy.dtype or Python type to cast entire pandas object to
+            the same type. Alternatively, use {col: dtype, ...}, where col is a
+            column label and dtype is a numpy.dtype or Python type to cast one
+            or more of the DataFrame's columns to column-specific types.
         raise_on_error : raise on invalid input
         kwargs : keyword arguments to pass on to the constructor
 
         Returns
         -------
         casted : type of caller
         """
-
-        mgr = self._data.astype(dtype=dtype, copy=copy,
-                                raise_on_error=raise_on_error, **kwargs)
-        return self._constructor(mgr).__finalize__(self)
+        if isinstance(dtype, collections.Mapping):
+            if self.ndim == 1:  # i.e. Series
+                if len(dtype) > 1 or list(dtype.keys())[0] != self.name:
+                    raise KeyError('Only the Series name can be used for '
+                                   'the key in Series dtype mappings.')
+                new_type = list(dtype.values())[0]
+                return self.astype(new_type, copy, raise_on_error, **kwargs)
+            elif self.ndim > 2:
+                raise NotImplementedError(
+                    'astype() only accepts a dtype arg of type dict when '
+                    'invoked on Series and DataFrames. A single dtype must be '
+                    'specified when invoked on a Panel.'
+                )
+
+            from pandas import concat
+            casted_cols = [self[col].astype(typ, copy=copy)
+                           for col, typ in dtype.items()]
+            other_col_labels = self.columns.difference(dtype.keys())
+            other_cols = [self[col].copy() if copy else self[col]
+                          for col in other_col_labels]
+            new_df = concat(casted_cols + other_cols, axis=1, copy=False)
+            return new_df.reindex(columns=self.columns, copy=False)
+
```

Comment:
this is going to raise if the columns are duplicates (not the dtype dict). Put them in order first (the original order). before passing to the concat. Further doing ``self[col]`` also will not work with duplicates. something like:

```
results = []
for c, col in self:
    if c in dtype:
      results.append( # cast col here)
    else:
     results.append(col)

new_df = concat(results, axis=1, copy=False)
...

```